students/Isabella_Kemp/lesson05/mailroom_part3 .py

```diff
@@ -0,0 +1,181 @@
+# Isabella Kemp
+# Mailroom part 3 - incorporating exceptions and comprehensions
+# 2/16/2020
+
+import sys
+import pathlib
+import time
+
+# List of donors and the amounts they have donated
+donors = {"John": [150080.00, 41.28],
+          "Irene": [1600.00, 24.47],
+          "Rob": [19000.00, 200.47],
+          "Kathy": [819.00, 34.5],
+          "Laureen": [830.00, 47.00, 982.13],
+          "Miles": [24.50, 87.00, 193.00]}
+
+main_prompt = ("\n\nWelcome to the Mailroom! Please choose an option from the menu:"
+               "\n1. Send a Thank you"
+               "\n2. Create a Report"
+               "\n3. Send letters to all donors"
+               "\n4. Quit"
+               "\n\n ----> ")
+
+
+# Common function to get input from user, call with a prompt string
+def get_user_input(prompt_string):
+    response = input(prompt_string)
+    return response
+
+
+# User can enter 'list' to get list of existing donors
+# User can enter 'listall' to get list of donors and their donations
+# User can enter existing donor name, new donation, donation is added to list
+# User can enter a new donor, new donation, user and donation is added to list
+# We finish with a thank you
+def thank_you():
+    while True:
+        name = get_user_input("Enter a donor name to send a thank you letter, "
+                              "'list' or 'listall' for list of donors, '4' will exit: ")
+        if name == "list":
+            # comprehension
+            donor_names = [key for key in donors]
+            print(donor_names)
+            # [print(key) for key in donors] For a prettier output, no None's are outputted
+        elif name == "listall":
+            # entire record name with donations
+            # comprehension, I realize this isnt appropriate with a print statement
+            # but it does not return None's and cleans up the lines, so I wanted to try it
+            [print("{0} {1}". format(key, donors[key])) for key in donors]
+        elif name == '4':  # allows them to quit and go back to menu
+            print("Finished processing thanks to donors...\n\n")
+            return
+        else:
+            # User entered a name, so process that
+            record = get_donor(name)
+            # print(record) # for debug comment out
+            if (record != None):
+                # donor is in our list,
+                # get donation amount, add donation to list, send thank you
+                print("user is in our list")
+                donation = get_donation()
+                record.append(donation)
+                donors[name] = record
+                email(name, donation)
+
+            else:
+                # new donor is not in our list, add the donor to our list
+                # get donation amount, add donation to list, send thank you
+                print("!!! NEW DONOR !!!")
+                donation = get_donation()
+                donors.update({name: [donation]})
+                email(name, donation)
```

Comment:
there is a lot of duplicate code above - note that this can be combined and reduce a lot of lines.
Let me know if you'd like to see specific code for this.